src/foam/mlang/mlang.js

```diff
@@ -3248,39 +3293,22 @@ foam.CLASS({
   methods: [
     {
       name: 'f',
-      code: function(o) {
-        if ( foam.core.Reference.isInstance(this.arg1) ) {
-          return o[property.name + '$find'].then(val => this.arg2.f(val));
-        }
-        return this.arg2.f(this.arg1.f(o));
+      code: async function(o) {
+        return o[arg2.name + '$find'];
       },
       javaCode: `
-        if ( getArg1() instanceof PropertyInfo && getArg2() instanceof PropertyInfo ) {
-          StringBuilder sb = new StringBuilder("find");
-          PropertyInfo p1 = (PropertyInfo) getArg1();
-          FObject obj1;
-          if ( p1 instanceof AbstractFObjectPropertyInfo ) {
-            Object val = getArg1().f(obj);
-            return val == null ? null : getArg2().f(val);
-          }
-          try {
-            obj1 = (FObject)obj.getClass().getMethod(StringUtil.capitalize(p1.getName()), foam.core.X.class).invoke(obj, ((FObject)obj).getX());
-          } catch ( Throwable t ) {
-            return null;
-          }
-          if ( obj1 == null ) return null;
-          try {
-            return getArg2().f(obj1);
-          } catch ( Throwable t ) {
-            Logger logger = (Logger) getX().get("logger");
-            if ( logger == null ) {
-              logger = new StdoutLogger();
-            }
-            logger.error(t);
-            return null;
+        PropertyInfo p1 = (PropertyInfo) getArg2();
+        FObject refObj = null;
+        try {
+          refObj = (FObject)getArg1().getObjClass().getMethod("find" + StringUtil.capitalize(p1.getName()), foam.core.X.class).invoke(obj, ((FObject)obj).getX());
```

Comment:
Can you check if Relationship has a helper method to do this, so we only have method name creation logic in one place, please. 
if it doesn't, please add.

Reply:
yes, there is, but it's not really clear to me where we could retrieve a relationship for a reference property. is there some kind of relationship dao? thanks

Reply:
my understanding of relationship is that we use them for java code generation. so I'm wondering if we maybe create something like AbstractReferencePropertyInfo and copy info from relationship to it?

Reply:
@olhabn ok, leave that for now and create another ticket to capture the support method - and ideas.